### PR TITLE
[OTLP exporter] Remove dependency on deprecated gRPC library

### LINKED ISSUE #3421: Use `Grpc.Net.Client` with `netstandard2.0` for `OpenTelemetry.Exporter.OpenTelemetryProtocol`
# Feature Request

**Is your feature request related to a problem?**

The old `Grpc` package copies `libgrpc_csharp_ext.x64.dylib` and `libgrpc_csharp_ext.x86.so` all over the place. That's ~17MB of unnecessary libs being copied to all of our the services that references `OpenTelemetry.Exporter.OpenTelemetryProtocol` in our solution. 

**Describe the solution you'd like:**

It should be possible to use `Grpc.Net.Client` with `netstandard2.0` instead. In fact, in theory, any `net462` or greater app should be able to reference `netstandard2.0`.

### LINKED ISSUE #4613: Exporter.OTLP on .NET Fx relays on deprecated package
# Feature Request

**Is your feature request related to a problem?**

[OpenTelemetry.Exporter.OpenTelemetryProtocol](https://github.com/open-telemetry/opentelemetry-dotnet/tree/5f825cb8fcadd3f5689467e314bb25fa9f1f8419/src/OpenTelemetry.Exporter.OpenTelemetryProtocol) for .NET Framework is referencing [GrpcCore](https://www.nuget.org/packages/Grpc.Core)
which is deprecated. There will be no security updates for this package: https://groups.google.com/g/grpc-io/c/OTj5mb1qzb0/m/ZNVo1S3iBAAJ

**Describe the solution you'd like:**

No dependencies on Grpc.Core.
Dependency on Grpc.Net.Client seems to be not feasibly. Based on error from additional context it is supported only on Win11+
https://learn.microsoft.com/en-us/aspnet/core/grpc/netstandard?view=aspnetcore-7.0#net-framework

**Describe alternatives you've considered.**

N/A

## Additional Context

Local test on .NET Framework 4.6.2 with Grpc.Net.Client.

```log
Actual:   System.PlatformNotSupportedException: gRPC requires extra configuration on .NET implementations that don't support gRPC over HTTP/2. An HTTP provider must be specified using GrpcChannelOptions.HttpHandler.The configured HTTP provider must either support HTTP/2 or be configured to use gRPC-Web. See https://aka.ms/aspnet/grpc/netstandard for details.
             at Grpc.Shared.HttpHandlerFactory.CreatePrimaryHandler()
             at Grpc.Net.Client.GrpcChannel.CreateInternalHttpInvoker(HttpMessageHandler handler)
             at Grpc.Net.Client.GrpcChannel..ctor(Uri address, GrpcChannelOptions channelOptions)
             at Grpc.Net.Client.GrpcChannel.ForAddress(Uri address, GrpcChannelOptions channelOptions)
             at OpenTelemetry.Exporter.OtlpExporterOptionsExtensions.CreateChannel(OtlpExporterOptions options) in C:\GitHub\opentelemetry-dotnet\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\OtlpExporterOptionsExtensions.cs:line 39
             at OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.ExportClient.OtlpGrpcTraceExportClient..ctor(OtlpExporterOptions options, TraceServiceClient traceServiceClient) in C:\GitHub\opentelemetry-dotnet\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\Implementation\ExportClient\OtlpGrpcTraceExportClient.cs:line 36
             at OpenTelemetry.Exporter.OtlpExporterOptionsExtensions.GetTraceExportClient(OtlpExporterOptions options) in C:\GitHub\opentelemetry-dotnet\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\OtlpExporterOptionsExtensions.cs:line 83
             at OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.OtlpExporterOptionsExtensionsTests.<>c__DisplayClass4_0.<GetTraceExportClient_GetClientForGrpcWithoutUnencryptedFlag_ThrowsException>b__0() in C:\GitHub\opentelemetry-dotnet\test\OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests\OtlpExporterOptionsExtensionsTests.cs:line 135
             at Xunit.Record.Exception(Func`1 testCode) in /_/src/xunit.core/Record.cs:line 47
   at OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.OtlpExporterOptionsExtensionsTests.GetTraceExportClient_GetClientForGrpcWithoutUnencryptedFlag_ThrowsException() in C:\GitHub\opentelemetry-dotnet\test\OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests\OtlpExporterOptionsExtensionsTests.cs:line 144

```

Consider as candidate for 1.6 release.

### PR DESCRIPTION
**Pull request by alanwest**
Fixes #3421
Fixes #4613 

We have discussed recently dropping our dependency on the old gRPC library. Let's use this PR to make a final decision.

The gRPC C# library may technically still be in "maintenance mode", though from the following comment this may change "any day now" https://github.com/grpc/grpc/issues/32719#issuecomment-1620399357:

> So in practice, Grpc.Core is still "maintained", but it might get declared deprecated any day now, without any extra notice. There's still some small chance that we'll reconsider (based on the data available to us) and continue supporting Grpc.Core for some more time, but AFAICT this seem unlikely and I definitely wouldn't count on it.

There is a possibility that dropping this dependency will be a breaking change for some .NET Framework users. The newer `Grpc.Net.Client` library does support .NET Framework [under certain scenarios](https://learn.microsoft.com/aspnet/core/grpc/supported-platforms?view=aspnetcore-7.0#net-grpc-client-requirements) (i.e., Windows 11 / Windows Server 2022 using the WinHttpHandler).

Until very recently, the newer library `Grpc.Net.Client` did not have a `net462` target. To support our .NET Framework users, this would have required us to take a dependency on `WinHttpHandler` and to construct the gRPC channel to use it. However, `Grpc.Net.Client` as of version 2.56.0 introduced a `net462` target which appears to make it "just work" on .NET Framework (still with the Win 11/Server 2022 requirement). See: https://github.com/grpc/grpc-dotnet/pull/2220#issuecomment-1662222421

---

I'm just putting this PR out there to start the conversation. I have not yet had a chance to try it out on .NET Framework to validate that using `Grpc.Net.Client` version 2.56.0 without any special configuration works.

### CODE REVIEW

**Comment by pengweiqhca** [comment]
`Grpc.Net.Client` has many [limitation](https://learn.microsoft.com/en-us/aspnet/core/grpc/supported-platforms?view=aspnetcore-7.0#net-grpc-client-requirements), I think should provide a [`Grpc.Core.Api.CallInvoker` or `Grpc.Core.ChannelBase`](https://github.com/open-telemetry/opentelemetry-dotnet/pull/4815#issuecomment-1719977560) factory, it will give the most flexibility with how a user provides their desired gRPC connection.

**Comment by Kielek** [comment]
Unsupported scenarios OTLP over GRPC on
* Windows 10 (EOL Oct 14, 2025), https://learn.microsoft.com/en-us/lifecycle/products/windows-10-home-and-pro
* Windows 2016 (EOL Jan 12, 2027) https://endoflife.date/windows-server
* Windows 2019  (EOL Jan 09. 2029) https://endoflife.date/windows-server

I was trying it on Win10 and it is really not working.

**Comment by alanwest** [comment]
> I think should provide a [`Grpc.Core.Api.CallInvoker` or `Grpc.Core.ChannelBase`](https://github.com/open-telemetry/opentelemetry-dotnet/pull/4815#issuecomment-1719977560) factory, it will give the most flexibility with how a user provides their desired gRPC connection.

Unfortunately, we want to avoid exposing either `CallInvoker` or `ChannelBase` in our options class as you've done in #4815. The reason is because this creates a coupling between our API and specific gRPC client libraries. We are currently considering if there is a way we can eliminate our dependency on any gRPC library entirely.

**Comment by alanwest** [comment]
> I was trying it on Win10 and it is really not working.

Yes, as I suspected, definitely a breaking change. We'll discuss in the SIG meeting tomorrow. Whether we drop it now or not we need to begin to consider what our plan is if, in the event the library becomes unmaintained, a security issue is found.

**Comment by pjanotti** [comment]
This may be a moment to bring #2959 into consideration.

**Comment by alanwest** [comment]
> This may be a moment to bring https://github.com/open-telemetry/opentelemetry-dotnet/issues/2959 into consideration.

This crossed my mind, however, changing the default to OTLP/HTTP would expand the impact to all users rather than just users running on older versions of Windows. Nonetheless, we'll discuss today.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OpenTelemetry.Exporter.OpenTelemetryProtocol.csproj [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>$(DefaultTargetFrameworks);netstandard2.1</TargetFrameworks>
+    <TargetFrameworks>$(DefaultTargetFrameworks)</TargetFrameworks>
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OpenTelemetry.Exporter.OpenTelemetryProtocol.csproj:3]: `<DefaultTargetFrameworks>net6.0;netstandard2.0;net462</DefaultTargetFrameworks>`

For quick/easy ref: https://github.com/open-telemetry/opentelemetry-dotnet/blob/7cb92d3b6dbad169048e5aa07939ef613f2587e2/build/Common.prod.props#L6

**Comment by cijothomas** [comment]
Historically, this repo has dropped support for certain .NET frameworks when the frameworks went out of support itself, and this was done with a minor version bump. This PR is similar to that : OTLP Exporter in .NET Framework needs a particular grpc library, and that library itself is going out of support. Its reasonable for OTel .NET to drop support (and do minor version bump only) as well.

(Similar is [lack of Prometheus exporter for .NET Framework](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.Prometheus.HttpListener#prometheus-exporter-httplistener-for-opentelemetry-net) that can be used in production. There is no underlying library support for that, and we are not building it ourselves...)

**Review by cijothomas — COMMENTED** [review]
Agree with this direction. Not marking as explicitly approved, as we need to update changelogs/readme etc. to reflect this.

**Comment by Kielek** [comment]
@alanwest, there is new announcement about Grpc.Core package. The maintenance mode is extended about at least one year from now (October 2024).

See https://groups.google.com/g/grpc-io/c/iEalUhV4VrU?pli=1

I suppose we can wait a bit before merging these changes.

**Comment by alanwest** [comment]
> The maintenance mode is extended about at least one year from now (October 2024).

Interesting, thanks for the heads up. Given the strong support expressed in removing the dependency, I'd still be a fan of removing it sooner than later giving folks more time to adjust.

I am moving this PR to draft for now. I have not had the time yet, but from our discussion in our SIG meeting a few weeks ago, my remaining TODO on this PR was to try adding an integration test for .NET Framework.

Even if we ultimately do decide to hold off removing the dependency awhile longer it'll still be nice to have this branch ready for whenever we are.

**Comment by cijothomas** [comment]
Tagging @alanwest

**Comment by Kielek** [comment]
I think that it was again started by @vishweshbankwar in #5731. I doubt that it will be finished by him. Do you know what is the status of this PR?

**Comment by reyang** [comment]
I think there is no plan to reopen/resurrect this PR. The plan is to remove gRPC library dependency entirely as mentioned by @Kielek https://github.com/open-telemetry/opentelemetry-dotnet/pull/4860#issuecomment-2356551723.

**Comment by alanwest** [comment]
I cannot find any updates regarding the end of support extension to October 2024. If we find that it will be extended even further then I do not think there's any rush to remove the dependency to the old gRPC library.

Though, if someone can validate that it truly is ending October 2024, then resurrecting this PR could make sense. It would in no way affect the longer term desire to eliminate dependencies on all gRPC libraries.

**Pull request closed without merging**